tools/checklocks/state.go

```diff
@@ -18,23 +18,53 @@ import (
 	"fmt"
 	"go/token"
 	"go/types"
+	"sort"
 	"strings"
 	"sync/atomic"
-	"sort"
 
 	"golang.org/x/tools/go/ssa"
 )
 
 // lockType represents the current state of the lock.
 // The 0 value represents the unlocked state.
+// The valid states are unlocked, locked, locked & deferredunlock
```

Comment:
It is not clear if there are 4 states or 3 from this. Is it just "unlocked, locked, locked & deferredunlock" or "unlocked, locked, deferredunlock, locked & deferredunlock"? "valid" can mean "user valid program" or "valid" within checklocks (can only occur due to a programming error in checklocks).

The comment "The valid states are unlocked, locked, locked & deferredunlock" reads as the states this can take on in the checklocks are "unlocked, locked, locked & deferredunlock".
The comment "If a state is only deferredunlock..." reads as "unlocked, locked, deferredunlock, locked & deferredunlock" are all possible.

Reply:
I think there should be 4 states in a user valid program `unlocked, locked, deferredunlock, locked & deferredunlocked`. The analysis should update the states based on the semantics of the program. I think `deferredunlock` is actually `unlocked & deferredunlock` since it is not locked. If it ends up outside of these 4 states, then it is a programming error within checklocks.